Centralise Ocean parameters

merijn opened the pull request:
Doesn't currently include validation because THCM parameters are passed via ocean and don't pass validation.

I think the best approach is to refactor THCM in a similar fashion so that Ocean can simply query the default parameters from THCM and extend its default parameters with those. The main obstacle is that some of THCM's parameters are only used in the constructor, so it's unclear what "setting" those would mean/look like. Specifically THCM parameters only used in the constructor to initialise other values are:

```
double qz      = paramList.get("Grid Stretching qz", 1.0);
int    itopo   = paramList.get("Topography", 1);
bool   flat    = paramList.get("Flat Bottom", false);
bool   rd_mask = paramList.get("Read Land Mask", false); //== false in experiment0
bool rd_spertm     = paramList.get("Read Salinity Perturbation Mask",false);
int coriolis_on    = paramList.get("Coriolis Force", 1);
int forcing_type   = paramList.get("Forcing Type", 0);
std::string spertm_file = paramList.get("Salinity Perturbation Mask", "no_mask_specified");
std::string windf_file = paramList.get("Wind Forcing Data", "wind/trtau.dat");

// Let THCM know the sst forcing file
std::string sst_file = paramList.get("Temperature Forcing Data", "levitus/new/t00an1");

// Let THCM know the sss forcing file
std::string sss_file = paramList.get("Salinity Forcing Data", "levitus/new/s00an1");

bool time_dep_forcing = paramList.get("Time Dependent Forcing",false);
int Nic = paramList.get("Integral row coordinate i", N-1);
int Mic = paramList.get("Integral row coordinate j", M-1);
```

@erik808 @Sbte what do you think should be done with those parameters?

Sbte commented:
- You forgot to remove the commit of the other pull request.
- The nice thing about `get` with default values is that it sets the parameter as `used`. See 
  https://trilinos.org/docs/dev/packages/teuchos/doc/html/classTeuchos_1_1ParameterList.html#acf06d867a960000f651b1d8657e083ff
  Using the templated `get` does not do this. Also using the parameter list as const implies that you do not do this (but knowing Trilinos, it's possible that they made it `mutable`, you could check this). The use of this `used` is that you can print the parameter list at the end of the run or initialization, and it tells you for every parameter whether you set it and it was used, the default value was used, or it was unused. It would be nice if we could retain this. Or get it to work properly if it doesn't work yet.
- I don't see a problem with setting parameters in THCM from Ocean only. To make sure you could put an assert in `THCM.C` to check that they were not changed at the end of the run. All initialization can be moved to Ocean. Unless @erik808 disagrees.
- Please use a `_` suffix for all class members. It's really confusing if you don't.

merijn commented:
> * The nice thing about `get` with default values is that it sets the parameter as `used`. See
>   https://trilinos.org/docs/dev/packages/teuchos/doc/html/classTeuchos_1_1ParameterList.html#acf06d867a960000f651b1d8657e083ff
>   Using the templated `get` does not do this. Also using the parameter list as const implies that you do not do this (but knowing Trilinos, it's possible that they made it `mutable`, you could check this). The use of this `used` is that you can print the parameter list at the end of the run or initialization, and it tells you for every parameter whether you set it and it was used, the default value was used, or it was unused. It would be nice if we could retain this. Or get it to work properly if it doesn't work yet.

If we want to make the original ParameterList the "single source of truth" we really don't want other parts of the code accidentally "adding" parameters that were not in the original list, because then we're back to square 1 of having no centralised places were all parameters can be found. I'll have to think how we can make that work together with still tracking whether a parameter was used or not.

Having a "default" parameter list we lose the ability to track whether a default was used anyway, since all the used parameters were already explicitly set in this default list I initialise from. If this is really important I think the simplest way to do it would be to subclass Teuchos::ParameterList and track "unused"/"used"/"default" in that?

The current THCM doesn't actually ever use the default value anyway, since the get function is only ever called for parameters that are actually in the list, see: https://github.com/nlesc-smcm/i-emic/pull/64/files#diff-5c096fb489dae5c82b49f6a8261e9a45L1902

Which basically means that it can only ever report used/unused. I actually think we don't lose anything here since there's no conditional selection of which variables to set, so each variable that THCM knows about is *always* set if present (so they can't be left unused). This means only variables that THCM doesn't know about will ever be "unused", but once we add the validation so that we only allow "known" parameters, that can't happen either.

Anyway, if you still think this functionality is important we can brainstorm some way to make it work again.

Sbte commented:
Adding parameters is not a problem. You should just always add them with the default value by using `get`. That way it will tell you it used the default value. For instance

```C++
#include <iostream>

#include "Teuchos_ParameterList.hpp"

int main()
{
    Teuchos::ParameterList pars;
    pars.set("Test 1", 1);
    pars.set("Test 2", 2);
    std::cout << pars.get("Test 1", 2) << std::endl;
    std::cout << pars.get("Test 3", 3) << std::endl;
    std::cout << pars.get("Test 3", 4) << std::endl;
    std::cout << pars;
}
```

gives as output

```
1
3
3
Test 1 = 1
Test 2 = 2   [unused]
Test 3 = 3   [default]
```

Note that the second `get` of "Test 3" is also 3, not 4! So in your `createDefaultParameterList` you should just pass the parameters that were passed to the constructor, and then use `get` instead of `set`.

The reason that the THCM parameters are not read in this way yet, is that the default values are in the Fortran code. If you move them to the C++ code this can be done with the proper get method.

merijn commented:
> Adding parameters is not a problem. You should just always add them with the default value by using get. That way it will tell you it used the default value. For instance

It's a problem in the sense that, if I query Ocean "hey, which parameters do you support?" and 'Test3' is not in the default list, then omuse doesn't know 'Test3' is a parameter that we can set, since omuse only knows about the parameters it can query. That's why I made the ParameterList const, that way we know it can only use parameters that were already in the list (and thus must be in the default list, since the validation will reject any parameter that's not in the default list).

Sbte commented:
Yes, but that's my point. You just `get` every parameter into the default parameter list, as well as any other parameter list. So `getDefaultParameters` becomes

```
Teuchos::ParameterList getDefaultParameters()
{
Teuchos::ParameterList params;
setParameters(params);
return params;
}
```

and `setParameters` only contains `get` methods. You can also call `setParameters` from the constructor to actually set the parameters.

merijn commented:
Just to be sure, are we talking about the same thing? I thought you were referring to the change in THCM.C's readParameters, but do you mean you're referring to the parameter setting inside Ocean?

Sbte commented:
Both. They should both do the same thing. `ReadParameters` should become `setParameters` for THCM.

merijn commented:
Relatedly, I've been using const for the ParameterList references being passed in, but if you want to use this used/unused/default setup that'd mean having `setParameters`/constructors mutating the argument `ParameterList`. I'm not a big fan of that, but if you think it's important I can switch to doing that.

Although I think the same information would be available in `params_`, so that might work without needlessly mutating the input lists...

merijn commented:
My plan was to first centralise the parameterlists, then ensure setParameters verifies against the default, and then figure out how to add the validators (I actually already have some stuff for validation, but it breaks everything until I sort out THCM and solver_params)

Sbte commented:
Ok, perfect. The commits that you pushed now also seem fine to me. The only thing I noticed is the `belosParamList_` instead of `belosParamList`.

merijn commented on src/ocean/THCM.C:
```diff
@@ -219,46 +207,27 @@ THCM::THCM(const Teuchos::ParameterList& params, Teuchos::RCP<Epetra_Comm> comm)
-    int    itopo   = paramList.get("Topography", 1);
-    bool   flat    = paramList.get("Flat Bottom", false);
-    bool   rd_mask = paramList.get("Read Land Mask", false); //== false in experiment0
+    double qz      = paramList.get<double>("Grid Stretching qz");
```
These and a couple of others are only used in the constructor to initialise parts of the Fortran code and/or initialise files (that are presumably read by the Fortran code). I wasn't sure if it was safe/sensible to allow setting them after the initial constructor creation?

If yes, then it's an easy fix. If no, then things get a bit messier/uglier...

Sbte replied:
Like I said, anything related to the grid, which this is I think, should only be used in the initialization. If you want to change them, you should reinitialize everything.

merijn replied:
You mean this [Fortran initialisation](https://github.com/nlesc-smcm/i-emic/blob/d347c4d72e0a9d82a965b38e8de2a93ea476a2c2/src/ocean/THCM.C#L341-L348), right? Is it that initialise function expensive to run? (i.e. would it be ok to run everytime `setParameters` is called?) And is it fine to just call again?

Sbte replied:
No, that is not doable. For instance during a continuation, you constantly want to adjust parameters, and every time you run the initialization, a new ocean will be created that is initialized to 0.

merijn replied:
Ok, then I think it'd actually be best to make a distinction between "default parameters for initialisation" and "other default parameters" (i.e. that can be tweaked at runtime), so we can distinguish the two and validate them properly.

merijn commented on src/ocean/THCM.C:
```diff
@@ -832,6 +795,16 @@ THCM::THCM(const Teuchos::ParameterList& params, Teuchos::RCP<Epetra_Comm> comm)
         col_scaling->PutScalar(1.0);
     }
 
+    double val;
```
This should really be part of `setParameters` but it has to run after the initialisation of the Fortran model or things crash. This ties in with the earlier comment. Ideally it should be possible to dummy initialise the Fortran code then set the right parameters later so this and the previous problem are fixed.

merijn commented on src/ocean/THCM.C:
```diff
@@ -2058,7 +2010,7 @@ bool THCM::setParameter(std::string label, double value)
     {
         // set monthly forcing data
-        bool time_dep_forcing = paramList.get("Time Dependent Forcing",false);
+        bool time_dep_forcing = paramList.get<bool>("Time Dependent Forcing");
```
I think this should probably become an instance variable.

merijn commented on src/ocean/THCM.C:
```diff
@@ -3017,3 +2969,138 @@ Teuchos::RCP<const Epetra_MultiVector> THCM::getNullSpace()
+    //Teuchos::ParameterList& startParams =
+    params.sublist("Starting Parameters");
+
+    /*
```
So, I used this to fill in the default parameters for THCM, turns out those **aren't** actually default, because initialising this way causes tests too fail. We should probably determine what defaults the fortran code **actually** initialises with...

merijn commented:
Some of the constructor only parameters used in THCM are used to initialise this [here](https://github.com/nlesc-smcm/i-emic/blob/f64cbe17cef473d96fbcf97af5d2974a1b6dfbdd/src/ocean/THCM.C#L317-L318), would it be sensible to allow updating these later or should these only really be set at initialisation?

The fortran code gets initialised [here](https://github.com/nlesc-smcm/i-emic/blob/master/src/ocean/THCM.C#L375-L384) but ideally we want to be able to set/overwrite these parameters later via `setParameters` is that feasible?

Sbte commented:
I think everything but the grid can be changed later. Maybe even the grid could be changed later, but then you would have to re-initialize most of the code anyway, so it would be better to just call the constructor again.

merijn commented:
Relatedly, right now I've been adding the same functions to different classes, but would it perhaps make sense to define a baseclass that provides virtual implementations of all those getParameters/setParameters function and inherit from those? (This would especially be useful to avoid duplicating a lot of work if/when we also add a separate `setInitParameters` to account for stuff that can only be set at initialisation time.

Sbte commented:
Yeah, go ahead.

merijn commented:
I actually just realised the inheritance approach is kinda pointless, since half the common code are class methods, so can't be inherited anyway...

merijn commented on src/ocean/THCM.C:
```diff
@@ -3018,3 +2353,764 @@ Teuchos::RCP<const Epetra_MultiVector> THCM::getNullSpace()
+
+    Teuchos::ParameterList& startParams = params.sublist("Starting Parameters");
+    /*
+    for (int i=0; i<= _NPAR_ + _NPAR_TRILI; i++)
```
The "default" values returned here by `defaultParameter` aren't, well, the actual defaults, because uncommenting this code breaks most of the tests. We should probably fix these default values, but I'm unsure where the current values are set.

Sbte replied:
They are in `usr.F90`

merijn commented:
hmm, I should probably just eliminate `setInitParameters` now that I'm done refactoring and move the body back into the constructor to shrink the diff.

merijn commented:
Ok, so the parameter initialisation works, but something is funky. I have the same failure as Travis on my machine, but it doesn't appear to be related to the parameter code I added, because the commit adding the updated `defaultParameter` code breaks **even when it's not used**. I added some debugging code and I noticed that after some changes (like, print formatting) in the debug code the tests suddenly start working again. The Fortran parameters I touch appear to have the same value in the working runs as well as the broken ones, so it seems that the `intt_ocean` and `reft_ocean` somehow read from memory that's either uninitialised or accidentally initialised in some cases. I spend all day yesterday trying to find out the cause, but so far no success.

merijn commented:
Ok, after spending ages trying to track it through the code I have nailed the issue down to the `AL_T` parameter. This parameter is initialised by the Fortran code to different values based on other parameters. Which conflicts with providing a default value in the parameterlist (well, unless I reverse engineer/duplicate the relevant Fortran code on the C side.

Do we ever need/want to set this parameter manually? If not, then the simplest solution would be to just drop it from the accepted parameterlist. The alternate approach is to only allow setting it later and relying on the default initialisation during the initialisation of THCM.

Sbte commented:
I think you have to move all the initialization to the C++ side to do this properly. At the moment the tests may not fail for the other parameters, but unfortunately we do not cover all relevant cases in the tests I think. The place where most of these parameters are set, including `AL_T` is `stpnt`. Maybe @erik808  has something useful to say about this.

erik808 commented:
So is it `AL_T` because it's the first of its 'kind', the first entry in the THCM parameter array? Do you get problems with `RAYL` when you drop `AL_T` from the accepted parameters? I don't see occurrences of setting `AL_T` other than in `stpnt `in `usrc.F90`. I guess the other entries may give the same problems. 

So yeah, moving all initialization to cpp would be best. 

In this case `AL_T` (stands for alpha_tau, wind stress coefficient) is not exactly fixed and I think we would want to do adjustments here to sync POP and I-EMIC.

merijn commented:
> In this case AL_T (stands for alpha_tau, wind stress coefficient) is not exactly fixed and I think we would want to do adjustments here to sync POP and I-EMIC.

Ok, then we need to handle this specially anyway. There's a similar problem [here](https://github.com/nlesc-smcm/i-emic/pull/64/files#diff-5c096fb489dae5c82b49f6a8261e9a45R667-R668) where the default values for the row and column parameter depend on other parameters.

I think there's basically 2 options:
1) Add an additional boolean parameter for each of these to indicate whether to use the "default" or use the provided value, or
2) Use ParameterList validators to enforce the right behaviour to compute the default dynamically when it's missing. On the other hand, this is a tricky because I don't immediately see a validator that is able to look at other parameters in the same ParameterList, so then computing the "default" might be tricky.

Sbte commented:
I think the problem here is that all parameters have to be set to their defaults from the omuse side. It wouldn't be a problem if we only write and read them when we actually need them. I feel like you're creating workarounds for something that shouldn't be a problem in the first place.

merijn commented:
> I think the problem here is that all parameters have to be set to their defaults from the omuse side. It wouldn't be a problem if we only write and read them when we actually need them. I feel like you're creating workarounds for something that shouldn't be a problem in the first place.

Ok, after diving into the Fortran code and thinking more about it, I think this can be fixed without changing much of the C++ or Fortran code and without introducing any hacks/workarounds/etc.

This would involve moving the parameter initialisation into C++. Basically, all the relevant initialisation happens in `vmix_par` (only called from `stpnt`), `stpnt` (only called from `init`) and `init` itself is only called from C++. By moving the logic of `vmix_par` and `stpnt` out of the Fortran code and into the C++ we can keep the current behaviour without any workarounds or funky business and still have everything nicely centralised in terms of the ParameterList, does that sound good? @Sbte @erik808

Sbte commented:
That's what I meant 2 comments ago (sorry if I wasn't clear), so yeah, that sounds fine by me.

erik808 commented:
I'm also ok with this. Hope this reduces the spaghetti related to vmix
flags as well

On Tue, Nov 19, 2019 at 6:09 PM Sven Baars <notifications@github.com> wrote:

> That's what I meant 2 comments ago (sorry if I wasn't clear), so yeah,
> that sounds fine by me.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nlesc-smcm/i-emic/pull/64?email_source=notifications&email_token=ADAZG7LCEYELCDEO5V7X67LQUQMT7A5CNFSM4JGI32O2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEO6W4I#issuecomment-555608945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADAZG7JHUQOWZYTXDTCCNFTQUQMT7ANCNFSM4JGI32OQ>
> .
>

merijn commented:
I take it back, not nearly as straightforward as I thought. The default for `AL_T` depends on `dz` and `dfzT(l)`, which are initialised by [`grid`](https://github.com/nlesc-smcm/i-emic/blob/master/src/ocean/grid.F90), I could duplicate that computation to compute `AL_T` in the C++ code, but that sorta thing always makes me paranoid, because I have no idea if there's any risk of someone modifying that Fortran initialisation of `grid`, leading to a mismatch between the computed default and what it'd be in Fortran.

merijn commented:
Given the amount of time disentangling `AL_T` from all the Fortran allocations and the initialisation of `dfzT` has already consumed I'm more considering one of my earlier suggestions to just special case it. All the other parameters can be easily handled, it's just `AL_T` that's difficult.

Sbte commented:
I still don't think the special case is the solution.

grepping for relevant quantities gives me

```
ocean/grid.F90:  dz = (zmax-zmin)/L
ocean/TRIOS_Domain.C:        double dz = (bounds[zmax]-bounds[zmin])/sizes[L];
ocean/global.F90:    dz = (zmax-zmin)/L
ocean/OceanGrid.C:    dz = (domain->Zmax()-domain->Zmin())/(domain->GlobalN());
ocean/grid.F90:     dfzT(k) = dfdz(ze(k),qz)
ocean/grid.F90:     ze(k)  = (real(k)-0.5)*dz + zmin
ocean/grid.F90:     dfdz = qz  / (tth * ch * ch)
ocean/grid.F90:     dfdz = 1.0 + (1.-qz)*(1.-2.*z)
```

Everything else comes from `init` (so you already have it at the C++ side) or is a local variable. So it seems doable. The only thing that makes it difficult is that there is duplication between the `global` and `usr` modules, including the usage of global variable names (which I really really hate). Take care that the one used for `stpnt` is the `usr` version. You don't have to implement the full `dfzT` array, just take `dfzT(l)` as a constant and add a comment where it comes from.

Eventually we may want to just get rid of the Fortran code, because all of those globals make it really confusing and they also make it impossible to use shared memory parallelization of any kind. If we have a sufficient amount of tests this should be doable. I already did this for similar other codes.

merijn commented on src/ocean/THCM.C:
```diff
@@ -3018,3 +2979,161 @@ Teuchos::RCP<const Epetra_MultiVector> THCM::getNullSpace()
+    int M = params.get<int>("Global Grid-Size m");
+
+    params.get("Integral row coordinate i", N-1);
+    params.get("Integral row coordinate j", M-1);
```
@Sbte Only problem with the idea of using validateParametersAndSetDefaults are these two lines, since their defaults depends on the values for the `Global Grid-Size`, which works in the current version, but can't really be done using set, I think?

The only option would be to set the default value equal to the default `Global Grid-Size`, but then any code with a different grid size needs to also explicitly set these.

merijn replied:
They actually don't seem to be explicitly set anywhere, except for `test_integrals` (where they're unused) and `test/global/ocean_params.xml` which doesn't appear to be used anywhere either...

merijn commented:
That gets rid off #65, the only thing in the way of `validateParametersAndSetDefaults` is those annoying `Integral row coordinate i` parameters I mentioned. They don't seem to be set by any of the tests, I think the options are to either:

1) Get rid of them as parameters
2) Use `-1` to indicate the default value and update after construction
3) I saw in the docs that there's a more general "parameter list acceptor", but which could handle them properly, but that seems a lot of work for little pay off.

merijn commented:
@Sbte Do you have any opinion/comment on the `Integral row coordinate i` parameter stuff?

Sbte commented:
It's something you can change to make sure extra conditions are imposed on something that is in the ocean, not on land. The comment in the xml says it should also always be on the last processor. In popular culture this point is usually referred to as "pirates". Maybe @erik808 has more to say about it.

merijn commented:
Ok, I think this one is basically done now. The macOS tests are failing due to new XCode release on Travis-CI, I will work on fixing those tomorrow and then this can be merged as far as I am concernced.

merijn commented:
And everything passes \o/

Sbte commented on src/ocean/Ocean.C:
```diff
@@ -975,19 +952,21 @@ void Ocean::initializeBelos()
 
     // Create Belos parameterlist
-    belosParamList_ = rcp(new Teuchos::ParameterList());
+    RCP<Teuchos::ParameterList> belosParamList_ = rcp(new Teuchos::ParameterList());
```
Please get rid of the underscore

Sbte commented on src/ocean/Ocean.C:
```diff
@@ -52,38 +52,23 @@ extern "C" _SUBROUTINE_(set_seaice_parameters)(SeaIce::CommPars*);
+    : Ocean(Comm, paramListPtr == Teuchos::null ? Teuchos::ParameterList() : *paramListPtr)
+{}
+
+Ocean::Ocean(RCP<Epetra_Comm> Comm, const Teuchos::ParameterList& oceanParamList)
```
You really want to pass a `Teuchos::RCP` here, see my comment for `setParameters`

Sbte commented on src/ocean/Ocean.C:
```diff
@@ -2200,27 +2179,80 @@ void Ocean::setPar(std::string const &parName, double value)
+const Teuchos::ParameterList& Ocean::getParameters()
+{ return params_; }
+
+void Ocean::setParameters(Teuchos::ParameterList newParams)
```
This should receive a reference to a parameterlist, not a copy. And then `validateParametersAndSetDefaults` should be called on that reference, not on a local copy. That way you can see in the parameterlist what gets used and what not (and when we properly document the parameters, that will be added too). The same should be done for the THCM parameterlist.

In `run_ocean`, if you do
```
    std::cout << *oceanParams << std::endl;
```
at the end, it should display
```
Load state = 0
Load salinity flux = 0
Save salinity flux = 1   [default]
Load temperature flux = 0   [default]
Save temperature flux = 1   [default]
Use legacy fort.3 output = 0   [default]
Use legacy fort.44 output = 1   [default]
Save column integral = 0   [default]
Max mask fixes = 5   [default]
Analyze Jacobian = 1   [default]
Input file = ocean_input.h5   [default]
Output file = ocean_output.h5   [default]
...
```
not
```
Load state = 0   [unused]
Load salinity flux = 0   [unused]
THCM -> 
 Problem Description = test   [unused]
 Global Bound xmin = 300   [unused]
...
```

You can use the `ocean` run directory and set
```
    <Parameter name="Global Grid-Size n" type="int" value="8"/>
    <Parameter name="Global Grid-Size m" type="int" value="8"/>
    <Parameter name="Global Grid-Size l" type="int" value="8"/>
```
if you want a quick example.

merijn replied:
Hmm...the reason for doing it this way was so that it would be valid to call `setParameters` with a subset of the parameters to update one or more parameters that `Ocean` is using.

If you do it your way you run into the problem calling `setParameters` with a parameter list that has a single parameter you want to update, then you reset all the other parameters back to default. You should still be able to figure out which parameters are used/defaulted by simply calling `getParameters()` and inspecting the results of that, no?

merijn replied:
hmm, I need to think about this for a bit, because that doesn't propagate the THCM usage back to Ocean, no. So I agree the constructors should be mutable references or `RCP`, but I need to think how to make that work with `setParameters` so that you can still update subsets of parameters.

Or, I guess the simpler way is to simply say that `setParameters` must always have *all* parameters and doesn't allow/do updates to subsets and then changing OMUSE to simply remember the previous parameters you set, which shouldn't be too difficult.

Sbte replied:
What about you do

```
    params_.setParameters(newParams);
```
at the end of `setParameters` and then do
```
Teuchos::ParameterList result(params_);
```
at the start of `getDefaultParameters`?

merijn replied:
I think the correct solution is a compromise. I think the current validation is the right way to implement it, since your comment would actually make `getDefaultParameters` equal to `getParameters`.

But you're right that the parameter lists should be `RCP` pointers, and I think I finally got why they were like that. (On the other hand, refactoring them to use const references was helpful in sorting out the rest of the code, so that's good!)

Sbte replied:
Ok, then maybe add an optional argument to `getDefault(Init)Parameters`? If it's set, you initialize result with that, if not then you do what is currently done.

merijn replied:
I don't think that's needed? In the new commit I've kept that code as is and it correctly tracks whether parameters use the default or not. I tried your example of having an `ocean_params.xml` that only sets the THCM dimensions and nothing else and when I print `oceanParams` at the end of `run_ocean` it correctly reports all variables as default, except the dimensions I set in THCM, which it reports as used.

Sbte commented on src/ocean/Ocean.H:
```diff
@@ -63,6 +63,7 @@ public:
     using LandMask  = Utils::MaskStruct;
 
 protected:
+    Teuchos::ParameterList params_;
```
This should probably be an RCP so the Belos parameterlist also sets the used and default parameters. Even better if this can be removed.

merijn replied:
I don't think it can be removed if we wanna support querying the current parameter values. On the other hand, we could "not support" that, I guess?

Sbte replied:
If you implement my previous suggestion (using params_ to initialize result), then we have to keep it anyway.

Sbte commented on src/ocean/Ocean.H:
```diff
@@ -168,6 +164,12 @@ public:
     Teuchos::RCP<Epetra_Comm> Comm() const { return comm_; }
 
+    const Teuchos::ParameterList& getParameters();
+    void setParameters(Teuchos::ParameterList);
```
Please be consistent and also add the name of the argument.

Sbte commented on src/ocean/Ocean.C:
```diff
@@ -95,8 +80,8 @@ Ocean::Ocean(RCP<Epetra_Comm> Comm, ParameterList oceanParamList)
     //  to THCM::Instance()
-    Teuchos::ParameterList &thcmList =
-        oceanParamList->sublist("THCM");
+    const Teuchos::ParameterList &thcmList =
```
This should not be const if you fix the other comments

merijn commented:
To summarise: I think your objection is that you want to see which parameters are used/defaulted in the argument to `setParameters`, yes? I think the problem is that you can't see which parameters are used anyway, as all the `get` operations going through `params_`, so you can only see which ones are used from the `ParameterList` you originally created `Ocean` with, or from the one you get via `getParameters`.

And if you initialise `Ocean` with some parameter list and then later use `setParameter` to update/set some subset of the parameters of `Ocean`, then you have no way (in the `setParameter` argument) to track which was were already set before, because if you use validate on the `setParameter` on the argument it will just default every missing parameter and overwrite the original `Ocean` config.

However, if you use the current approach of having an immutable argument to `setParameter` then you can simply update the stored `params_` member and revalidate it and then you'll still correctly have both the defaulting and unused information in the original `Ocean` parameter list, as well as in the parameter list return by `getParameters`.

merijn commented:
The current approach also correctly propagates updates to, say, THCM back to the parameterlist in Ocean.

Sbte commented:
I pushed something to https://github.com/merijn/i-emic/compare/params...Sbte:params. The same would have to be done for THCM. Does this work for you?

merijn commented:
The downside of the approach in that PR is that if you call `setParameters` on THCM, it will update the THCM parameters and mark things as used/default in the argument to `setParameters`, but the parameter list in `Ocean` will still report the old parameters THCM was initialised with?

The advantage of my original approach is that for all the places where a sublist initialises some other class (like THCM/solver and Ocean), then updating one will also update the original. Because the THCM `RCP` is just a reference to the original `Ocean` sublist.

Sbte commented:
Can that be solved by initializing THCM before the parameters? And by calling the THCM version of `setParameters` only through the Ocean version of `setParameters`?

In your version I think only the initial parameterlist was updated with default/used values, which is not what we want.

merijn commented:
> Can that be solved by initializing THCM before the parameters? And by calling the THCM version of setParameters only through the Ocean version of setParameters?

Maybe? That's what I am thinking about now.

> In your version I think only the initial parameterlist was updated with default/used values, which is not what we want.

The idea of my version was that `params_` becomes the source of truth. It's initialised with the original `RCP` pointer (which is then shared to THCM so the lifetime of both is correct), and `validateAndSetDefaults` will correctly report variables that are not in the default list. It will also correctly mark any missing variables as defaulted.

Then, if you later call `setParameters` with a subset of the parameters, it will update those in `params_`, which resets the default flag (if present) and then revalidates the list, so values that are updated later will also be correctly reported as not default when you call `getParameters`. The original ParameterList `RCP` will also reflect this, so both printing out oceanParams in `run_ocean` and `getParameters()` will correctly produce this information.

So it should still be trivial to report which parameters were used/defaulted, which was the main goal, right?

merijn commented:
Big brain time! I think I know how to do both.

merijn commented:
@Sbte this version uses my version of the code (so changes to THCM, etc. are all correctly tracked in `Ocean`), but is changed so that if you call `setParameters` the `ParameterList` correctly represents which values are defaulted and which were set. Are you ok with that approach?

I'm not so happy with the "empty" parameterlist in the ocean constructor right now, but it's time to go home, so I'll find a nice way to get rid of it after the weekend.

Sbte commented:
I still don't like this because you set the parameters to your "single source of truth" before you actually validate them.

Since it should never happen that anyone other than Ocean has access to THCM, you may just share `params_` between Ocean and THCM so they stay in sync. I don't really like this either, however.

I think they both need to care about their own parameter list. Ocean should not care about the THCM parameters. If you want to make sure `getParameters` returns the right thing, just make it call `getParameters` from THCM.

ipelupessy commented:
Hi Sven,

 I am trying to disentangle to above discussion...if I understand correctly merijns state of affairs of the PR is that Ocean and THCM classes use references to a *single* instance of the parameters list (THCM a reference to the appropiate sublist of the Ocean/ input parameters list). The change you propose then introduce an extra copy at both Ocean and THCM level?? is this a correct summary of the above? I am a bit lost in the various comments, so can you talk me through on why this is necessary? (merijn is trying to convince me it is not!)

Sbte commented:
The common point I try to make with all my comments on this draft is that we want to be able to get from printing the parameter list what parameters were used (from the input list, and the defaults), and the documentation (so we don't need this across all xml files). This is nicely implemented in Trilinos and after merging this pull request, documentation can easily be added by adding an extra argument to the call where the defaults are set.

However, Merijn wants to have a single source of truth parameter list, which is shared with all classes. This is not necessary, however. I also think it complicates things, because if we change parameter handling in one class in the future, we have to be careful to not break other classes that share the parameter list. I also think that from a code design point of view, it is nicer if THCM only cares about THCM parameters, and Ocean only cares about Ocean parameters. Ocean should not have to be careful with making sure it doesn't accidentally overwrite THCM parameters.

Also, since this is apparently complicated, please add tests.

ipelupessy commented:
hmm, from what I understand (please correct me if I am wrong) THCM is decoupled from Ocean in so far that it gets a reference on instantiation (which happens to be a sublist)..the sublist is handled transparently in THCM. Ocean could mess this up, but this involves changes to the sublist..which one wouldn't do accidently??

Sbte commented:
Well, this pull request actually changes the THCM sublist from Ocean here (if the lists are shared):

https://github.com/nlesc-smcm/i-emic/pull/64/files#diff-24e73cc71ccdf36ee75c0b8350cdf86cR2231

ipelupessy commented:
ok, I am not sure 100% what setparams of teuchos::ParameterList actually does, but my understanding is that this snippet:
```
params_->setParameters(newParams);
params_->validateParametersAndSetDefaults(getDefaultInitParameters());
```
first updates params_ from the newParams and then validates them, checks whether any parameters are in there that should not be and fills in any missing parameters from defaults?
so question is what happens/ should happen if:
1) newParams contains THCM sublist
2) when it doesn't
I think at option 1: it should just propagate the sublist (and probably it does?) 
not sure about option 2?

can you two (@merij, @Sbte) comment on this?

Sbte commented:
Well, as I commented, the first thing that is wrong is already that it updates the parameters BEFORE validating that they are valid. But as for your comments, what happens:

1. Since `params_` is the "single source of truth" the THCM parameter list gets set as well (but the actual parameters do not!!!).
2. Nothing

What I think should happen is that `setParameters` also calls `thcm->setParameters` and sets the internal parameter list from there. It is safe to call setParameters without a THCM sublist (or without any parameters at all for that matter).

ipelupessy commented:
> Well, as I commented, the first thing that is wrong is already that it updates the parameters >BEFORE validating that they are valid. But as for your comments, what happens:
ok, from what I get the newparams can also be just a subset or one particular parameter that you want to change - in this case validating before will fill it with default values for the other parameters??

> 
>     1. Since `params_` is the "single source of truth" the THCM parameter list gets set as well (but the actual parameters do not!!!).
> 
>     2. Nothing
> 
> 
> What I think should happen is that `setParameters` also calls `thcm->setParameters` and sets the >internal parameter list from there. It is safe to call setParameters without a THCM sublist (or without >any parameters at all for that matter).

ok, i see... this is necessary...basically setParameters and constructor are the only places that should be used to set/change parameters..its not too far from the current setup I think though..let me think about this (and also @merijn to comment)...

ipelupessy commented:
this commit fixes the validation and propagates the sublist to THCM setParameters (and fixes constructor). the paramlist passed to the constructor will get updated, which is ok. The only thing which you should not do is update the paramlist afterwards with the expectation that these changes will be committed to the sim code..

Sbte commented:
Apparently that made the tests fail...

Also with that change the parameter list passed to setParameters doesn't get updated with what is done with the parameters (used, unused, default, etc). The entire point of passing a nonconst reference to that method is that the parameter list gets updated.

I still don't see what's wrong with my approach, since at the very least it's much easier to read.

ipelupessy commented:
yea we are working on it (its not entirely obvious why it fails..

ipelupessy commented:
this should fix it.

note that at the end the input newparam gets updated so that you have same information as your proposal. in your proposal the original input parameterlist to the constructor is only updated after the constructor - not with later changes. I guess what is preferable depends a bit on the usage pattern!

The main items for conversation are according to @merijn:
i) should the classes have their own private copy (currently: no, you: yes)
ii) should the argument to setparameters and the constructor argument be updated (yes, yes)
iii) should the original argument to the constructor be kept up to date after later setparameters (yes, no)

ipelupessy commented:
btw we are not fundamentally opposed to your proposal (merijn is ok with implementing that if you wish so), but I think we should first reach agreement on how it looks on the "user side" cause I got the feeling Merijn was chasing a moving target and led to a bit of unhappiness ;-)

merijn commented:
> Also, as for the extra thing you added to setParameters: Erik noted to me earlier today that we have to make sure (test) that this happens correctly for all parameters. I suggested during one of our earlier meetings that all parameter setting (and handling of that) is moved to the setParameters function (and removed from the constructor). Merijn noted that some parameters are only allowed to be set during the construction, so these may be an exception (but less exceptions is better).

Ideally I think the check we added in `setParameters` should be deleted (there and in the constructor) and replaced with an error during validation.

As for testing, I didn't want to invest any time into writing any tests until we at least agreed on the actual API provided by `Ocean` & friends.

Sbte commented:
Sure, maybe that check should be deleted, but there are many more things that are done in he constructor that depend on parameters that can be set. If the getting these parameters has been moved to `setParameters`, the stuff that depends on it also has to be moved. But I guess this can wait until after this has been merged, so it doesn't get even bigger.

Let's say the ultimate goal (not from this pull request, but after many more pull requests) is to be able to call the constructor first, then `setParameters`, and then some `Initialize` function (to call the fortran `init` function and handle all stuff that should be immutable through calling `setParameters` only).

Sbte commented:
> i) should the classes have their own private copy (currently: no, you: yes)
> iii) should the original argument to the constructor be kept up to date after later setparameters (yes, no)

On my bike I came up with an example of why this is preferable. If the parameter list is shared, we get this:

Say we have a class C with internal parameters a_ and b_ that can be set through the parameter list with "A" and "B".

```C++
Teuchos::RCP<Teuchos::ParameterList> params = Teuchos::rcp(new Teuchos::ParameterList);
params->set("A", 1);
params->set("B", 2);

C c(params);
// internal parameter a_ = 1
// internal parameter b_ = 2

// lots of code

params->set("A", 3);

// lots of code

// internal parameter a_ = 1
// internal parameter b_ = 2
Teuchos::ParameterList newParams;
newParams.set("B", 4);
c.setParameters(newParams);
// internal parameter a_ = 3!!!!!
// internal parameter b_ = 4
```

Now a_ is something completely different while we didn't pass it to setParameters. Note that this doesn't even have to happen in the same file. It can be changed from somewhere completely different because of the shared RCP.

ipelupessy commented:
ok, that can indeed be a problem; in effect the input parameterlist is only safe for reading in our case..@merijn shall we change it and go for @Sbte approach? note that this will make the "check the parameters" part (after constructor or setparameters) brittle in the sense that the input parameter list only provides a valid view immediately after the call..but maybe it is better to explicitly to a getParameters call on the class if you want an up to date status...

ipelupessy commented:
ok, I have changed parameter list now..(let me know what you think, @Sbte ..it may need some cleanup/simplfication)

Sbte commented on src/ocean/Ocean.C:
```diff
@@ -202,7 +179,7 @@ Ocean::Ocean(RCP<Epetra_Comm> Comm, ParameterList oceanParamList)
         Teuchos::rcp(new Epetra_Import(*sIndexMap_, state_->Map()));
 
-    INFO(*oceanParamList);
+    INFO(oceanParamList);
```
This should be changed back now

ipelupessy replied:
ok

Sbte commented on src/ocean/Ocean.H:
```diff
@@ -158,8 +153,9 @@ protected:
-          ParameterList oceanParamList);
+    Ocean(Teuchos::RCP<Epetra_Comm> Comm);
+    Ocean(Teuchos::RCP<Epetra_Comm> Comm, Model::ParameterList oceanParamList);
+    Ocean(Teuchos::RCP<Epetra_Comm> Comm, const Teuchos::ParameterList& oceanParamList);
```
The implementation only has 2 constructors.

ipelupessy replied:
ok2

ipelupessy commented:
to do, tests to:

  - [ ] checking that the defaults are set
  - [ ] used and unused are set
  - [ ] we can't validate with an incorrect parameter 
  - [ ] changing the parameter list after the constructor doesn't change the internal parameter list.

ipelupessy commented:
please add any other to do items!

Sbte commented:
I made an issue for the remaining stuff that is required after this pull request. I also updated your comment with something else I came up with.

merijn commented:
Rebasing these changes into a cleaner/more minimal PR.
